### PR TITLE
Split SourceABC to multiple interfaces

### PR DESCRIPTION
**Pull request by m-novikov**
This pull request splits SourceABС into multiple interfaces. Old SourceABC was used in Image, Slice and Data sources.
But they all had different function and different interfaces. To better communicate intent now there is
`DataSourceABC` for querying "raw" data
`Slice2DSourceABC` for slicing data sources and producing 2d arrays.
`ImageSourceABC` for converting 2d arrays into QImage

### CODE REVIEW

File: tests/test_utility/test_abc.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    def test_type_error_on_undefined_abstract_method(self):
+
+        with pytest.raises(TypeError):
+            self.IObject()
+
+    def test_abstract_signal_raises_type_error(self):
+        with pytest.raises(TypeError):
+            self.ISignal()
+
+    def test_abstract_signal_raises_type_error(self):
```
Comment (k-dominik) [inline tests/test_utility/test_abc.py]: is this some pytest dark magic that makes multiple methods with the same name work? In any case, I'd prefer not to have that.

File: volumina/pixelpipeline/datasources/interface.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    because its underlying datasource is in an indeterminate state.
+    In such cases, the requester should simply ignore the error.
+    The datasource has the responsibility of sending a dirty notification
+    when the source is ready again.
+    """
+
+    pass
+
+
+class DataRequestABC(ABC):
```
Comment (k-dominik) [inline volumina/pixelpipeline/datasources/interface.py]: Is there a good reason why this does not reside in `pixelpipeline.asyncabcs`?

File: tests/test_pixelpipeline/test_slicesources.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    def testRequest(self):
+        self.ss.setThrough(0, 1)
+        self.ss.setThrough(2, 2)
+        self.ss.setThrough(1, 127)
+
+        sl = self.ss.request((slice(None), slice(None))).wait()
+        self.assertTrue(np.all(sl == self.raw[1, :, :, 127, 2]))
+
+        sl_bounded = self.ss.request((slice(0, 3), slice(1, None))).wait()
+        self.assertTrue(np.all(sl_bounded == self.raw[1, 0:3, 1:, 127, 2]))
```
Comment (k-dominik) [inline tests/test_pixelpipeline/test_slicesources.py]:
```suggestion
        numpy.testing.assert_array_equal(sl_bounded, self.raw[1, 0:3, 1:, 127, 2])
```

File: volumina/pixelpipeline/slicesources.py [hunk 4]
```diff
@@ -42,7 +43,7 @@
 logger = logging.getLogger(__name__)
 
 
-class SliceRequest(object):
+class SliceRequest(RequestABC):
```
Comment (k-dominik) [inline volumina/pixelpipeline/slicesources.py]:
```suggestion
class PlanarSliceRequest(RequestABC):
```

File: volumina/pixelpipeline/imagepump.py [hunk 5]
```diff
@@ -351,15 +351,15 @@ def __init__(self, layerStackModel, sliceProjection, sync_along=(0, 1, 2)):
 
     # mappings
     def layerToSliceSources(self, layer):
```
Comment (k-dominik) [inline volumina/pixelpipeline/imagepump.py]: don't know if this should be renamed to be uber consistent

File: volumina/pixelpipeline/datasources/interface.py (hunk 2 above)
Comment (m-novikov) [reply volumina/pixelpipeline/datasources/interface.py]: Some requests (e.g. GrayscaleImageRequest) don't implement getResult method.

File: volumina/pixelpipeline/imagepump.py (hunk 5 above)
Comment (m-novikov) [reply volumina/pixelpipeline/imagepump.py]: renamed `layerToPlanarSliceSources`

File: volumina/pixelpipeline/datasources/interface.py (hunk 2 above)
Comment (k-dominik) [reply volumina/pixelpipeline/datasources/interface.py]: sure, I mean the whole interface - i'd say it qualifies as async?

Comment (m-novikov) [reply volumina/pixelpipeline/datasources/interface.py]: I am not even sure that `asyncabc` is a reasonable name :) It should be also `interfaces.py` or something

**Comment by m-novikov** [comment]
Moved interfaces of pixelpipeline to interface
Removed getResult method.
* results are already cached for every wait method
* calling getResult on not awaited request will return None, which is error-prone as we have temporal coupling here.

**Comment by k-dominik** [comment]
> Removed getResult method.

great

**Pull request merged**